### DISCUSSION TITLE
Support for multiple projects within a workspace

### DISCUSSION THREAD

**Discussion by we-wake-act21** [eclipse-langium/langium#1340] [Q&A] [answered]
I am working on incorporating support for multiple projects within a single workspace, ensuring that the scopes of one project remain isolated from others.

Currently, a project is identified by the location of a specific project file (`project.project`).

In the provided figure, you can observe that `new_test_project` and `project` are distinct projects coexisting within the same workspace named `Examples`.

![image](https://github.com/eclipse-langium/langium/assets/116638890/f1e669a3-9b58-4a20-846c-e7f7aeec5a77)

To achieve this, I am seeking guidance on the most effective approach for implementation.

**Comment by msujew** [comment, marked as answer]
Hey @we-wake-act21,

this should be relatively straightforward, assuming you have a good way of identifying which file belongs to a given workspace. You will need to override the global scope computation:

https://github.com/eclipse-langium/langium/blob/728b57982b782b31b88ea97aa2a1d3cfe57208b0/packages/langium/src/references/scope-provider.ts#L100-L102

in a custom scope provider and return a dedicated, filtered scope based on the current project. You can find the currently scoped document via the `context` argument.

Note that the current scope provider + index manager is heavily optimized towards a single global scope. You might run into performance issues, if you don't take care of stuff like caching etc.
